set-balance.md

```diff
@@ -168,6 +192,346 @@ A `Result` is considered an `Action`.
+
+    syntax AccountLock ::= lock ( id: LockID, until: Int, amount: Int, reasons: Set )
+ // ---------------------------------------------------------------------------------
+
+    syntax Bool ::= activeLock (AccountLock, Int, WithdrawReason, Int      ) [function]
+                  | activeLocks(Set,         Int, WithdrawReason, Int      ) [function]
+                  | activeLocks(List,        Int, WithdrawReason, Int, Bool) [function, klabel(activeLocksAux)]
+ // -----------------------------------------------------------------------------------------------------------
+    rule activeLock(AL, NOW, REASON, BALANCE) => NOW <Int until(AL) andBool BALANCE <Int amount(AL) andBool REASON in reasons(AL)
+
+    rule activeLocks(ALS, NOW, REASON, BALANCE) => activeLocks(Set2List(ALS), NOW, REASON, BALANCE, false)
+
+    rule activeLocks(.List, _, _, _, RESULT) => RESULT
+    rule activeLocks((ListItem(AL) => .List) REST, NOW, REASON, BALANCE, RESULT => RESULT orBool activeLock(AL, NOW, REASON, BALANCE))
+```
+
+Slashing and repatriation of reserved balances
+==============================================
+
+The first of these is also used by `slash`.
+
+* `slash_reserved`
+* `repatriate_reserved`
+
+```k
+    syntax Action ::= "slash_reserved" "(" AccountId "," Int ")"
+ // ------------------------------------------------------------
+    rule [slash-reserved]:
+         <k> slash_reserved(ACCOUNT, AMOUNT)
+          => set_reserved_balance(ACCOUNT, maxInt(0, RESERVED_BALANCE -Int AMOUNT))
+         ...
+         </k>
+         <accounts>
+           <account>
+             <accountID> ACCOUNT </accountID>
+             <reservedBalance> RESERVED_BALANCE </reservedBalance>
+             ...
+           </account>
+         </accounts>
+         <totalIssuance> TOTAL_ISSUANCE => TOTAL_ISSUANCE -Int minInt(RESERVED_BALANCE, AMOUNT) </totalIssuance>
+
+    syntax Action ::= "repatriate_reserved" "(" AccountId "," AccountId "," Int ")"
+ // -------------------------------------------------------------------------------
+    rule [repatriate-reserved]:
+         <k> repatriate_reserved(SLASHED, BENEFICIARY, AMOUNT)
+          => set_free_balance(BENEFICIARY, BENEFICIARY_FREE_BALANCE +Int minInt(SLASHED_RESERVED_BALANCE, AMOUNT))
+          ~> set_reserved_balance(SLASHED, SLASHED_RESERVED_BALANCE -Int minInt(SLASHED_RESERVED_BALANCE, AMOUNT))
+         ...
+         </k>
+         <accounts>
+           <account>
+             <accountID> SLASHED </accountID>
+             <reservedBalance> SLASHED_RESERVED_BALANCE </reservedBalance>
+             ...
+           </account>
+           <account>
+             <accountID> BENEFICIARY </accountID>
+             <reservedBalance> BENEFICIARY_RESERVED_BALANCE </reservedBalance>
+             <freeBalance> BENEFICIARY_FREE_BALANCE </freeBalance>
+             ...
+           </account>
+         </accounts>
+      requires BENEFICIARY_FREE_BALANCE +Int BENEFICIARY_RESERVED_BALANCE >Int 0
```

Comment:
Should add a side-condition `SLASHED =/=K BENEFICIARY`.

Also, can it ever be the case that `SLASHED ==K BENEFICIARY`? If so, add a rule for that.

Reply:
The Rust code does not check for this case, and it can legitimately happen, IMO.

Reply:
Right, but the K you've written implicitely assumes that `SLASHED =/=K BENEFICIARY`, because they are distinct keys in the same `Map`.

So if the Rust code allows for both cases, you'll need to add the side condition to this rule and add another rule where they are the same (and only one account is touched).

Reply:
What happens if `SLASHED ==K BENEFICIARY` with the current rules?

Reply:
The existing rule will not apply.